substratevm/src/com.oracle.svm.reflect/src/com/oracle/svm/reflect/serialize/hosted/SerializationFeature.java

```diff
@@ -96,8 +117,150 @@ public void duringSetup(DuringSetupAccess a) {
+        try {
+            Field memberField = clazz.getDeclaredField(fieldName);
+            memberField.setAccessible(true);
+            Object object = memberField.get(constantObject);
+            Method getDeclaringClass = object.getClass().getDeclaredMethod("getDeclaringClass");
+            getDeclaringClass.setAccessible(true);
+            return (Class<?>) getDeclaringClass.invoke(object);
+        } catch (NoSuchFieldException | IllegalAccessException | NoSuchMethodException | InvocationTargetException exception) {
+            return null;
+        }
+    }
+
+    private static Class<?> getLambdaClassFromConstantObject(Class<?> clazz, Object constantObject) {
+        try {
+            Field instanceClassField = clazz.getDeclaredField("staticBase");
+            instanceClassField.setAccessible(true);
+            return (Class<?>) instanceClassField.get(constantObject);
+        } catch (IllegalAccessException | NoSuchFieldException ignored) {
+            return null;
+        }
+    }
+
+    /**
+     * It is not possible to avoid reflection when trying to get lambda class from
+     * {@link ConstantNode} instance. Lambda class is wrapped in the {@code ConstantNode#value}
+     * which is of type {@code DirectHotSpotObjectConstantImpl}. That class is protected and it is
+     * not visible in the {@link SerializationFeature}. The only way to access fields and methods of
+     * this class is through the reflection. There are 2 different ways of getting lambda from the
+     * {@code ConstantNode#value}:
+     *
+     * 1. For capturing classes from the JDK {@code ConstantNode#value} of type
+     * {@code DirectHotSpotObjectConstantImpl} has {@code DirectHotSpotObjectConstantImpl#object} of
+     * type {@code DirectMethodHandle$Constructor} on Java 17 and in previous versions. Lambda class
+     * can be extracted from {@code DirectMethodHandle$Constructor#initMethod} using the
+     * {@code MemberName#getDeclaringClass()} or from {@code DirectMethodHandle#member} using the
+     * {@code MemberName#getDeclaringClass()}. Since all of those classes are protected and are not
+     * in the same package as the {@link SerializationFeature}, the only way to get the lambda class
+     * is using the reflection.
+     *
+     * 2. For capturing classes that are user-defined classes, {@code ConstantNode#value} of type
+     * {@code DirectHotSpotObjectConstantImpl} has {@code DirectHotSpotObjectConstantImpl#object} of
+     * type {@code DirectMethodHandle$StaticAccessor} type. In that case, lambda class is stored in
+     * the {@code DirectMethodHandle$StaticAccessor#staticBase}. Since
+     * {@code DirectMethodHandle$StaticAccessor} is also protected class and cannot be imported in
+     * the {@link SerializationFeature}, it is not possible to get values from this field in any
+     * ways other than the reflection.
```

Comment:
It doesn't seem plausible that this is the only way. Instead, it should be possible to get the declaring class of a constructor method handle by calling `type().returnType()` on it.

Reply:
I've tried using this. But the only thing I can get from this is capturing class of a lambda, not really lambda class. Avoiding reflection to get the object (comment above) was successful, but here I don't really get anything valuable. Before I push other changes, I wanted to consult with you on this topic. Do you have any other ideas?
Just to be sure that I understood idea, I got `DirectMethodHandle$Constructor` using `SnippetReflectionProvider.asObject` and then called `type().returnType()` on it.

Reply:
`type.returnType()` does not return what we expected it would. But I used some other reflection providers to get all the relevant information and to get lambdas from registered capturing classes.